console/console-init/ui/src/modules/iot-device/components/AddGateways/AddGateway.tsx

```diff
@@ -6,125 +6,333 @@
   GridItem,
-  Grid
+  Grid,
+  Popover,
+  DropdownPosition
 } from "@patternfly/react-core";
-import { DeviceListAlert } from "modules/iot-device/components";
-import { deviceIDRegExp } from "types/Configs";
+import { InfoCircleIcon } from "@patternfly/react-icons";
+import { DropdownWithToggle, TypeAheadSelect } from "components";
+import "./pf-overrides.css";
+import { gatewayTypeOptions } from "modules/iot-device/utils";
 
 export interface IGatewaysProps {
   header?: string;
-  gateways?: string[];
-  returnGateways?: (gateways: string[]) => void;
+  gatewayDevices?: string[];
+  gatewayGroups?: string[];
+  returnGatewayDevices?: (gateways: string[]) => void;
+  returnGatewayGroups?: (gatewayGroups: string[]) => void;
 }
 
 export const AddGateways: React.FunctionComponent<IGatewaysProps> = ({
   header = "Add permitted devices as gateways to this new device",
-  gateways: gatewayList = [],
-  returnGateways
+  gatewayDevices: deviceList = [],
+  gatewayGroups: groupList = [],
+  returnGatewayDevices,
+  returnGatewayGroups
 }) => {
-  const [inputID, setInputID] = useState<string>();
-  const [isValid, setIsValid] = useState<boolean>(false);
-  const [gateways, setGateways] = useState<string[]>(gatewayList);
+  const [gatewayDevices, setGatewayDevices] = useState<string[]>(deviceList);
+  const [gatewayGroups, setGatewayGroups] = useState<string[]>(groupList);
+  const [inputType, setInputType] = useState<string>("device_id");
+  const [selectedGroups, setSelectedGroups] = useState<any[]>([]);
+  const [selectedIDs, setSelectedIDs] = useState<any[]>([]);
 
   useEffect(() => {
-    returnGateways && returnGateways(gateways);
-  }, [gateways]);
+    returnGatewayDevices && returnGatewayDevices(gatewayDevices);
+  }, [gatewayDevices]);
 
-  const addGateway = () => {
-    if (inputID) {
-      setGateways([...gateways, inputID]);
-      setInputID("");
+  useEffect(() => {
+    returnGatewayGroups && returnGatewayGroups(gatewayGroups);
+  }, [gatewayGroups]);
+
+  const addGateway = async () => {
+    if (selectedIDs?.length) {
+      const gatewayIDs: string[] = [];
+
+      selectedIDs.forEach(device => {
+        !gatewayDevices.includes(device) && gatewayIDs.push(device);
+      });
+
+      setGatewayDevices([...gatewayDevices, ...gatewayIDs]);
+      setSelectedIDs([]);
+    }
+    if (selectedGroups?.length) {
+      const groupNames: string[] = [];
+
+      selectedGroups.forEach(group => {
+        !gatewayGroups.includes(group) && groupNames.push(group);
+      });
+
+      setGatewayGroups([...gatewayGroups, ...groupNames]);
+      setSelectedGroups([]);
+    }
+  };
+
+  const removeDevice = (id: string) => {
+    const idIndex = gatewayDevices.indexOf(id);
+    idIndex >= 0 && gatewayDevices.splice(idIndex, 1);
+    setGatewayDevices([...gatewayDevices]);
+  };
+
+  const removeGatewayGroup = (group: string) => {
+    const idIndex = gatewayGroups.indexOf(group);
+    idIndex >= 0 && gatewayGroups.splice(idIndex, 1);
```

Comment:
please move this code in if check so that we can avoid to unnecessary update state and re-rendering.

if(idIndex >= 0){
gatewayGroups.splice(idIndex, 1);
setGatewayGroups([...gatewayGroups]);
}